node/modules/storageminer.go

```diff
@@ -357,8 +357,9 @@ func NewProviderPieceStore(lc fx.Lifecycle, ds dtypes.MetadataDS) (dtypes.Provid
 	return ps, nil
 }
 
-func StagingMultiDatastore(lc fx.Lifecycle, r repo.LockedRepo) (dtypes.StagingMultiDstore, error) {
-	ds, err := r.Datastore("/staging")
+func StagingMultiDatastore(lc fx.Lifecycle, mctx helpers.MetricsCtx, r repo.LockedRepo) (dtypes.StagingMultiDstore, error) {
+	ctx := helpers.LifecycleCtx(mctx, lc)
+	ds, err := r.Datastore(ctx, "/staging")
```

Comment:
Doc for the `Datastore` says that ctx is only for startup, here we are passing a lifecycle context from start to the end of Lotus runtime.

Reply:
The `Datastore()` constructor should not retain the context. It should only be used for the start operation, and this is the only context that we have that is cancelled if the startup is cancelled.

Reply:
Hmm, we should be using the context from OnStart here, but it makes everything way harder.

Reply:
Yeah, that would be unnecessarily complex IMO.